---
title: "Pscan-R Examples"
output: html_document
date: "2022-09-30"
author: "Federico Zambelli"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**1. Building Background for Homo sapiens - 500bp upstream - UCSC hg38**

  **1.1. Getting Annotations**

```{r}
library("GenomicFeatures")
library("GenomicRanges")
library("BSgenome")
library("BSgenome.Hsapiens.UCSC.hg38")
library("BSgenome.Mmusculus.UCSC.mm10")
library("txdbmaker")
library("BSDA")
library("BiocParallel")

#a <- sapply(list.files(path = "./R", full.names = TRUE), source, echo = FALSE) #PROVVISORIO
#rm(a) #PROVVISORIO
#load_all() #PROVVISORIO

txdb <- makeTxDbFromUCSC(genome="hg38", tablename="ncbiRefSeqCurated") #import gtf annotation from UCSC
#txdb <- makeTxDbFromUCSC(genome="hg38", tablename="refGene") #import gtf annotation from UCSC
seqlevels(txdb) <- seqlevels(txdb)[1:24] #use only annotations on canonical chromosomes

prom_rng <- promoters(txdb, upstream = 500, downstream = 0, use.names = TRUE)

prom_seq <- getSeq(x = BSgenome.Hsapiens.UCSC.hg38, prom_rng) #promoter sequences
```

  **1.2. Getting Jaspar Profiles**
  
```{r}
library("TFBSTools")
library("JASPAR2022")
library("JASPAR2024") #doesn't work?
library("RSQLite") #needed for Jaspar2024

opts <- list()
opts[["collection"]] <- "CORE"
opts[["tax_group"]] <- "vertebrates"

J2022 <- getMatrixSet(JASPAR2022, opts) #core Jaspar 2022 profiles for vertebrates
#J2024 <- getMatrixSet(JASPAR2024, opts) #core Jaspar 2024 profiles for vertebrates
```
  
  **1.3. Building Pscan Background**
```{r}
source("../R/Build_Background.R")
source("../R/Helper_functions.R")
source("../R/Scan_and_post_processing.R")
source("../R/PSMatrix_class.R")

J2020_PSBG <- ps_build_bg(prom_seq, J2020, BPPARAM = MulticoreParam(24)) #Build Pscan Background

#ps_write_bg_to_file(J2020_PSBG, "J2020_hg38_500u_0d_UCSC.psbg.txt")

#rm(list = c("J2020", "opts", "prom_seq", "prom_rng"))
```

**2. pscan analysis**

```{r}
library("Biostrings")

prom_rng$tx_name_clean <- sub("\\..*$", "", prom_rng$tx_name)
target_prom_rng <- prom_rng[prom_rng$tx_name_clean %in% target[,1]]
prom_seq <- getSeq(x = BSgenome.Hsapiens.UCSC.hg38, target_prom_rng) #promoter sequences

#2020_PSBG <- ps_build_bg_from_file("--/BG_scripts/J2020_hg38_200u_50d_UCSC.psbg.txt", J2020) #if you need it from a file

results <- pscan(prom_seq, 2020_PSBG)

table <- ps_results_table(results)

saveRDS(table, "results.rds")

z_table <- ps_z_table(results)

#rm(prom_seq)
#rm(results)
#rm(table)
```
  ---
