---
title: "Pscan-R Examples"
output: html_document
date: "2022-09-30"
author: "Federico Zambelli"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**1. Building Background for Homo sapiens - 500bp upstream - UCSC hg38**

  **1.1. Getting Annotations**

```{r}
library("GenomicFeatures")
library("GenomicRanges")
library("BSgenome")
library("BSgenome.Hsapiens.UCSC.hg38")
library("BSgenome.Mmusculus.UCSC.mm10")

#a <- sapply(list.files(path = "./R", full.names = TRUE), source, echo = FALSE) #PROVVISORIO
#rm(a) #PROVVISORIO
load_all() #PROVVISORIO

#txdb <- makeTxDbFromUCSC(genome="hg38", tablename="ncbiRefSeqCurated") #import gtf annotation from UCSC
txdb <- makeTxDbFromUCSC(genome="hg38", tablename="refGene") #import gtf annotation from UCSC
seqlevels(txdb) <- seqlevels(txdb)[1:24] #use only annotations on canonical chromosomes

?s

prom_seq <- getSeq(x = BSgenome.Hsapiens.UCSC.hg38, prom_rng) #promoter sequences
```

  **1.2. Getting Jaspar Profiles**
  
```{r}
library("TFBSTools")
library("JASPAR2022")
library("JASPAR2024") #doesn't work?
library("RSQLite") #needed for Jaspar2024

opts <- list()
opts[["collection"]] <- "CORE"
opts[["tax_group"]] <- "vertebrates"

J2022 <- getMatrixSet(JASPAR2022, opts) #core Jaspar 2022 profiles for vertebrates
#J2024 <- getMatrixSet(JASPAR2024, opts) #core Jaspar 2024 profiles for vertebrates
```
  
  **1.3. Building Pscan Background**
  
```{r}

```
  
  ---
