---
title: "How to Generate a Background Model for PscanR algorithm"
author: "Diana Betelli"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{2. How to Generate a Background Model for PscanR algorithm}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 80
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

***Pscan*** is a software tool designed to analyze motif enrichment in a set of
sequences, such as promoters from co-regulated or co-expressed genes. It scans
these sequences using motifs that represent the binding specificity of known
transcription factors, identifying those that are significantly over or under
represented. This provides valuable insights into potential transcription
factors that may regulate the studied genes.

In order to assess the significance of the obtained result, a background model
must be computed. In other methods this is performed by shuffling the columns of
the motif or by building random sequence sets of the same size and length of the
sequence set investigated.

In Pscan, instead, the input sequences are considered a sample drawn from a
universe consisting of all promoter sequences available for the species under
investigation. This *universe* serves as the background model for statistical
analyses, ensuring a comprehensive and unbiased assessment of the analysis
outcomes.

# Installation

Before using Pscan, ensure that all required packages are installed. If not
already installed, they can be installed using `BiocManager::install()`.

```{r eval=FALSE, include=TRUE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install(c("txdbmaker", "GenomicFeatures", "GenomeInfoDb", 
                       "Biostrings", "TFBSTools", "PscanR", "BiocParallel"))

```

Then, load the necessary libraries:

```{r eval=FALSE, include=TRUE}
library(txdbmaker)
library(GenomicFeatures)
library(GenomeInfoDb)
library(Biostrings)
library(TFBSTools)
library(PscanR)
library(BiocParallel)
```

# Precomputed Background Models

For convenience, several background models have already been computed and
optimized to meet common research needs. The package includes precomputed
backgrounds for the following species:

-   *Homo sapiens* (both hg38 and hs1 assembly)

-   *Mus musculus* (both mm10 and mm39)

-   *Arabidopsis thaliana* (TAIR9)

-   *Drosophila melanogaster* (dm6)

-   *Saccharomyces cerevisiae* (sacCer3).

## Investigated Promoter Regions

The promoter regions analyzed in relation to the transcription start site (TSS)
include:

-   200 bp upstream and 50 bp downstream

-   450 bp upstream and 50 bp downstream

-   500 bp upstream and 0 bp downstream

-   950 bp upstream and 50 bp downstream

-   1000 bp upstream and 0 bp downstream

The motif matrices used to generate the background scores were obtained from the
JASPAR databases for the 2020, 2022, and 2024 releases.

To use a precomputed background, refer to the documentation for the
\`generate_psmatrixlist_from_background\` function.

# Generating Custom Backgrounds

To provide flexibility and allow users to generate custom backgrounds specific
to their experimental conditions, this document outlines the general pipeline
for generating a background using R. The necessary functions, packages, and
steps are detailed below.

## Import gene annotations

The first step is defining the species of interest and its genomic annotations.
Gene annotations provide essential information about the structure and function
of genes, including their location, the intron-exon structure, and regulatory
elements. These annotations are crucial for downstream analysis, such as
promoter regions definition and transcription factor binding site
identification.

The easiest way to retrieve annotation data is through UCSC using
`txdbmaker::makeTxDbFromUCSC`, which allow direct import of genome data. You can
modify the ***genome*** and ***tablename*** parameters based on your specie of
interest. It is convenient to use the ***txdbmaker*** package to create a
transcript database (TxDb) object. This serves as the foundation for defining
the promoter regions.

An example with *Homo sapiens* is reported below:

```{r eval=FALSE, include=TRUE}
txdb <- txdbmaker::makeTxDbFromUCSC(genome="hg38", 
                                    tablename="ncbiRefSeqCurated")
```

Alternatively, if the species is not present in UCSC, you can use a **GFF or GTF
annotation file** from public repositories like **Ensembl, NCBI, or TAIR**.
Example with *A. thaliana:*

```{r eval=FALSE, include=TRUE}
txdb <- txdbmaker::makeTxDbFromGFF('<path_to_your_file>', format = "gff3", dataSource="TAIR9", organism="Arabidopsis thaliana")
```

Another example is the most recent human genome assembly, hs1: if you are
interested in constructing a background for this assembly, please follow the
guidelines at <https://github.com/Bioconductor/txdbmaker/issues/1>

Since we typically want to focus on main chromosomes, we can filter them
accordingly:

```{r eval=FALSE, include=TRUE}
GenomeInfoDb::seqlevels(txdb) <- GenomeInfoDb::seqlevels(txdb)[1:24]

# Adjust [1:N] for your species 
```

## Define Promoter Regions

Promoter regions are DNA sequences located near the transcription start site
(TSS) of a gene. They play a crucial role in the regulation of gene expression
by serving as binding sites for transcription factors and RNA polymerase.
Typically, promoter regions extend both upstream and slightly downstream of the
TSS.

In bioinformatics, we can define promoter regions computationally using genomic
annotation tools. For instance, in R, the `GenomicFeatures` package provides a
straightforward way to extract promoter regions from a transcript database
(`txdb`). The following command retrieves promoter sequences by selecting a
region that extends **200 base pairs upstream** and **50 base pairs downstream**
of the TSS:

```{r eval=FALSE, include=TRUE}
# Define promoter regions

prom_rng <- GenomicFeatures::promoters(txdb, upstream = 200, downstream = 50, use.names = TRUE)
```

To save time on `TxDb` object creation, we have prepared a file containing the
promoter regions in advance.

```{r}
promoters_path <- system.file("extdata", "promoters.rds", 
                         package = "PscanR")
prom_rng <- readRDS(promoters_path)
prom_rng
```

Once the promoter regions are defined, the next step is to extract the
corresponding sequences from a genome reference. The reference genome used for
retrieving promoter sequences with `Biostrings::getSeq()` must be the same as
the one used to construct the transcript database (`txdb`) with
`makeTxDbFromUCSC()`. For example, if `txdb` was created using `"hg38"` as the
genome version, the corresponding `BSgenome` package
(`BSgenome.Hsapiens.UCSC.hg38`) should be used to ensure consistency.

```{r eval=FALSE, include=TRUE}
if (!requireNamespace("BSgenome.Hsapiens.UCSC.hg38", quietly = TRUE))
    install.packages("BSgenome.Hsapiens.UCSC.hg38")
library("BSgenome.Hsapiens.UCSC.hg38")
```

```{r}
# Retrieve promoter sequences

prom_seq <- Biostrings::getSeq(x = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38, prom_rng)
```

## Retrieve Transcription Factors Binding Site Profiles

To analyze binding sites, we retrieve **transcription factor binding matrices
(TFBMs)**, which represent nucleotide frequencies at each position in the
alignment.

Several curated databases provide collections of these matrices, including
**JASPAR** and **TRANSFAC**. These databases are regularly updated with new
experimental data, improving the accuracy of motif prediction.

Example of retrieving motif matrices from **JASPAR 2022**:

```{r eval=FALSE, include=TRUE}
if (!requireNamespace("JASPAR2022", quietly = TRUE))
    install.packages("JASPAR2022")
library("JASPAR2022")
```

```{r}
opts <- list()
opts[["collection"]] <- "CORE"  # The desired collection
opts[["tax_group"]] <- "vertebrates"  # The taxonomic group
JASPAR_matrices <- TFBSTools::getMatrixSet(JASPAR2022::JASPAR2022, opts)
```

By adjusting the collection type or taxonomic group, you can tailor the
selection to your specific research focus.

### Retrieving motif matrices from JASPAR 2024

*Note:* The usage of **JASPAR 2024** differs slightly from previous releases.
Please refer to the JASPAR 2024 documentation for additional details. Below is
an example of how to retrieve transcription factor binding matrices from JASPAR
2024:

```{r eval=FALSE, include=TRUE}
if (!requireNamespace("JASPAR2024", quietly = TRUE))
    install.packages("JASPAR2024")
library(JASPAR2024)
```

```{r}
# Disable SSL verification if necessary
httr::set_config(httr::config(ssl_verifypeer = 0L))

# Connect to the JASPAR 2024 database
JASPAR2024 <- JASPAR2024::JASPAR2024()
JASPARConnect <- RSQLite::dbConnect(RSQLite::SQLite(), JASPAR2024::db(JASPAR2024))

# Retrieve transcription factor binding matrices
J2024 <- TFBSTools::getMatrixSet(JASPARConnect, opts)
```

## Build the Pscan Background

With promoter sequences and transcription factor matrices ready, generate the
Pscan background using `ps_build_bg()`.

### Parallel Computation Considerations

The `ps_build_bg()` function supports parallel computation to enhance
performance. However, the number of cores used should be adjusted based on the
available hardware. If running on a local machine with limited computational
power, it is recommended to use only **one core**.

**Warning**: Using **a single core** (`BPPARAM = BiocParallel::SnowParam(1)`)
will result in **extremely slow computation**, especially with large datasets.
On a standard PC, the process may take **an entire day** to complete. To
significantly reduce computation time, it is strongly recommended to run the
analysis on a server with multiple available cores and adjust `BPPARAM`
accordingly.

In the following example, only 10 sequences are selected to reduce the
computational time.

For all the systems:

```{r echo=TRUE}
J2022_PSBG <- PscanR::ps_build_bg(prom_seq[1:10], JASPAR_matrices, BPPARAM = BiocParallel::SnowParam(1))

# Example for the first matrix
J2022_PSBG[[1]]
```

For Unix-like systems:

```{r eval=FALSE, include=TRUE}
# Build the Pscan background
J2022_PSBG <- PscanR::ps_build_bg(prom_seq[1:10], JASPAR_matrices, BPPARAM = BiocParallel::MulticoreParam(1))
```

Refer to the 'Comparing execution time with varying processors' vignette to go
into details of multithreading.

## Exporting the Generated Background

You can either directly use the generated background or save the constructed
background to a file for future use.

```{r eval=FALSE, include=TRUE}
# Export the background to a file 
PscanR::ps_write_bg_to_file(J2022_PSBG, "J2022_mm10_450u_50d_UCSC_curated.psbg.txt")
```

# The full background

When dealing with multiple analysis on the same background dataset, it can be
convenient to store an enlarged version of the background containing key metrics
such as the score, position and oligonucleotide sequence for each hit within the
background, to reduce computational time of the pscan algorithm on the set of
promoter sequences of interest.

When you export the background PSMatrixList with `ps_write_bg_to_file(),` this
information gets discarded to conserve memory. In fact, after the computation of
the background mean and standard deviation, the other background data are no
more needed since can be recomputed by `pscan()` (just for the subset of
promoter region of interest).

Although `pscan()` supports parallel computation, making the recalculation
relatively fast, scenarios that involve repeated analyses of the same promoter
region can benefit significantly from a more efficient approach. By preserving
the computed background information, the need for redundant calculations is
eliminated, leading to enhanced computational efficiency, especially when
working on large datasets or performing numerous iterations of the same
analysis.

## full-background Generation

To create a full-background dataset, you need to store the full output of the
`ps_build_bg()` function without discarding any computed values. This can be
fulfilled using the `save()` function in R and by setting the fullBG flag on
TRUE for `ps_build_bg()`. This will generate an .RData file containing the hits
score, position, strand, and oligonucleotide sequence, for each sequence scanned
with the PWMs.

## Application

Once the full-background has been created, it can be used in subsequent analyses
to streamline computations. The `pscan_full_bg()` function requires two inputs:

1.  **A vector of transcript names** -- This vector should contain the
    identifiers of the transcripts of interest.

2.  **The generated full-background dataset** -- The `.RData` file created in
    the previous step.

By using `pscan_full_bg()` with the saved full-background, the analysis process
becomes significantly more efficient, reducing computational overhead while
maintaining accuracy and reproducibility.

```{r}
# Display session information
sessionInfo()
```
