---
title: "Introduction to PscanR"
author: "Diana Betelli"
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to PscanR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

**Package overview:**

PscanR is an R implementation of the algorithm developed by 
Zambelli et all. (2009).
It is designed to analyze sequence motif enrichment, helping users to identify 
transcription factors (TFs) that are likely regulators of a set of co-regulated 
or co-expressed genes. By scanning sequences (e.g., promoters) against known 
TFs binding motif, PscanR assesses which motif are significantly over- or under-
represented and highlights candidate binding sites. 


**Workflow overview:**

***Input***: A set of sequences (e.g., promoters) and a library of known 
transcription factors binding motif (e.g., PWMs from JASPAR database).

***Analyses***: Identification of enriched motif and calculation of statistical 
significance (Z-scores and P-Values).

***Output***: A ranked list of enriched motif with associated P-Values and 
Z-scores.  

## Installation

To install and load the package from Bioconductor, follow these steps:

# Install Bioconductor (if necessary)
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

# Install the package from Bioconductor
BiocManager::install("PscanR")

# Load the package
library(PscanR)


```{r}
opts <- list()
opts[["collection"]] <- "CORE" # Core motif collection
opts[["tax_group"]] <- "vertebrates" # Vertebrate motifs
J2020 <- TFBSTools::getMatrixSet(JASPAR2020::JASPAR2020, opts) # Download motif

file_path <- system.file("extdata", "example_file_nfkb100.txt", 
                         package = "PscanR")
BG_path <- system.file("extdata", "J2020_hg38_200u_50d_UCSC.psbg.txt", 
                       package = "PscanR")
# Get the target regions
target <- read.csv(file_path, header = FALSE)

#txdb <- txdbmaker::makeTxDbFromUCSC(genome="hg38", 
#                                    tablename="ncbiRefSeqCurated")

txdb_path <- system.file("extdata", "txdb.db", package = "PscanR")
txdb <- AnnotationDbi::loadDb(txdb_path)
seqlevels(txdb) <- seqlevels(txdb)[1:24]

prom_rng <- GenomicFeatures::promoters(txdb, upstream = 200, downstream = 50, use.names = TRUE)
prom_rng$tx_name_clean <- sub("\\..*$", "", prom_rng$tx_name)
target_prom_rng <- prom_rng[prom_rng$tx_name_clean %in% target[,1]]

# Get the target sequences 
prom_seq <- Biostrings::getSeq(
  x = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38, 
  target_prom_rng) 

# Retrieve the background
J2020_PSBG <- PscanR::ps_retrieve_bg_from_file(BG_path, J2020)

# Run PscanR
results <- PscanR::pscan(prom_seq, J2020_PSBG, 
                 BPPARAM = BiocParallel::SnowParam(1)) 
# Use MulticoreParam on Unix systems. 

# Create the table of results
table <- PscanR::ps_results_table(results) 
head(table)

# Display session information
sessionInfo()
```

**Interpretation of results:**

***Z-score***: Indicates the degree of motif enrichment. A higher score suggests 
stronger enrichment.

***P-Value***: Provides the statistical significance of the results. Smaller values
indicates more significant results. 
