---
title: "PscanR Examples"
output: html_document
date: "2022-09-30"
author: "Federico Zambelli"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**1. Building Background for Homo sapiens - 500bp upstream the TSS - UCSC hg38**

  **1.1. Getting Annotations**

```{r}
# Import GTF annotation from UCSC
txdb <- txdbmaker::makeTxDbFromUCSC(genome="hg38", 
                                    tablename="ncbiRefSeqCurated")

# Alternatively use tablename="refGene"

# Use only annotations on canonical chromosomes 
GenomeInfoDb::seqlevels(txdb) <- GenomeInfoDb::seqlevels(txdb)[1:24] 

# Get the promoter regions
prom_rng <- GenomicFeatures::promoters(txdb, upstream = 500, 
                                       downstream = 0, 
                                       use.names = TRUE)

# Get the promoter sequences
prom_seq <- Biostrings::getSeq(
  x = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38, 
  prom_rng) 
```

  **1.2. Getting Jaspar Profiles**
  
```{r}
# Set configuration options for getMatrixSet() function
opts <- list()
opts[["collection"]] <- "CORE"
opts[["tax_group"]] <- "vertebrates"

# Core Jaspar 2020 profiles for vertebrates 
J2020 <- TFBSTools::getMatrixSet(JASPAR2020::JASPAR2020, opts)
```
  
  **1.3. Building Pscan Background**
```{r}
# Build PSCAN background 
J2020_PSBG <- PscanR::ps_build_bg(prom_seq, J2020, BPPARAM = BiocParallel::SnowParam(1))

# Use MulticoreParam() on Unix and adjust the number of workers based on how many you can use

# Save the result into a file
# ps_write_bg_to_file(J2020_PSBG, "BG_scripts/J2020_hg38_500u_0d_UCSC.psbg.txt")  

```

**2. pscan analysis**

```{r}
# Get the sequence identifiers from the pscan dataset
file_path <- system.file("extdata", "liver.txt", 
                         package = "PscanR")
target <- read.csv(file_path, header = FALSE) 

# Get the target promoter regions
prom_rng$tx_name_clean <- sub("\\..*$", "", prom_rng$tx_name)
target_prom_rng <- prom_rng[prom_rng$tx_name_clean %in% target[,1]]

# Get the target promoter sequences 
prom_seq <- Biostrings::getSeq(x = BSgenome.Hsapiens.UCSC.hg38, target_prom_rng) 

# Optionally, you can build the background from pre-computed files. 
# See 'How to Generate a Background Model for PscanR algorithm' for 
# further information.
# J2020_PSBG_precomputed <- generate_psmatrixlist_from_background('jaspar2020',
#                                                                 'hs',
#                                                                 c(-500,+0),
#                                                                 'hg38')
#
# Launch the PSCAN algorithm 
results <- PscanR::pscan(prom_seq, J2020_PSBG, 
                 BPPARAM = BiocParallel::SnowParam(1)) 
# Use MulticoreParam() on Unix-like systems 

# Generate the result table 
table <- PscanR::ps_results_table(results)

# Obtain the z-score matrix
PscanR::z_table <- ps_z_table(results)

```

