---
title: "PscanR Examples"
output: html_document
date: "2022-09-30"
author: "Federico Zambelli"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**1. Building Background for Homo sapiens - 500bp upstream - UCSC hg38**

  **1.1. Getting Annotations**

```{r}
# Only for development 
library(devtools)
load_all()

#a <- sapply(list.files(path = "./R", full.names = TRUE), source, echo = FALSE) 
#rm(a) #PROVVISORIO

# Import GTF annotation from UCSC
txdb <- makeTxDbFromUCSC(genome="hg38", tablename="ncbiRefSeqCurated") 

# Alternatively use tablename="refGene"

# Use only annotations on canonical chromosomes 
seqlevels(txdb) <- seqlevels(txdb)[1:24] 

# Get the promoter regions
prom_rng <- promoters(txdb, upstream = 500, downstream = 0, use.names = TRUE)

# Get the promoter sequences
prom_seq <- getSeq(x = BSgenome.Hsapiens.UCSC.hg38, prom_rng) 
```

  **1.2. Getting Jaspar Profiles**
  
```{r}
#library("JASPAR2024") #doesn't work?
#library("RSQLite") #needed for Jaspar2024

# Set configuration options for getMatrixSet() function
opts <- list()
opts[["collection"]] <- "CORE"
opts[["tax_group"]] <- "vertebrates"

# Core Jaspar 2020 profiles for vertebrates 
J2020 <- getMatrixSet(JASPAR2020, opts) 

# Core Jaspar 2022 profiles for vertebrates
#J2020 <- getMatrixSet(JASPAR2022, opts) 

# Core Jaspar 2024 profiles for vertebrates
#J2024 <- getMatrixSet(JASPAR2024, opts) 
```
  
  **1.3. Building Pscan Background**
```{r}
#source("R/Build_Background.R") # Not needed if you use load_all()
#source("R/Helper_functions.R")
#source("R/Scan_and_post_processing.R")
#source("R/PSMatrix_class.R")

# Build PSCAN background 
J2020_PSBG <- ps_build_bg(prom_seq, J2020, BPPARAM = MulticoreParam(24)) 

# Save the result into a file
#ps_write_bg_to_file(J2020_PSBG, "BG_scripts/J2020_hg38_500u_0d_UCSC.psbg.txt")  

#rm(list = c("J2020", "opts", "prom_seq", "prom_rng"))
```

**2. pscan analysis**

```{r}
# Get the sequence identifiers from the pscan dataset
file_path <- system.file("extdata", "example_file_nfkb100.txt", package = "PscanR")
target <- read.csv(file_path, header = FALSE) 

# Get the target promoter regions
prom_rng$tx_name_clean <- sub("\\..*$", "", prom_rng$tx_name)
target_prom_rng <- prom_rng[prom_rng$tx_name_clean %in% target[,1]]

# Get the target promoter sequences 
prom_seq <- getSeq(x = BSgenome.Hsapiens.UCSC.hg38, target_prom_rng) 

# Build the background from a file
#J2020_PSBG <- ps_build_bg_from_file(
#  "BG_scripts/J2020_hg38_500u_0d_UCSC.psbg.txt", 
#  J2020
#) 

# Launch the PSCAN algorithm 
results <- pscan(prom_seq, J2020_PSBG, BPPARAM = BiocParallel::MulticoreParam(1)) 
# Use SnowParam() insted of MulticoreParam() on windows

# Generate the result table 
table <- ps_results_table(results)

saveRDS(table, "results.rds")

# Obtain the z-score matrix
z_table <- ps_z_table(results)

```

