<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Build background matrices in parallel — ps_build_bg • PscanR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Build background matrices in parallel — ps_build_bg"><meta property="og:description" content="This function generates a background probability profile matrix list,
used to assess if a transcription factor binding motif is appearing more
often than by chance in the pscan() input sequences, using the BiocParallel
framework for parallel execution.
It computes background scores for each motif matrix, collected, for example,
from the JASPAR database, by scanning them against a set of regulatory
sequences (e.g., gene promoters) using the `ps_scan` function.
Note that the promoter region analysed may vary."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PscanR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Background_generation.html">How to Generate a Background Model for PscanR algorithm</a>
    </li>
    <li>
      <a href="../articles/Execution_time_cores.html">Comparing execution time with varying processors</a>
    </li>
    <li>
      <a href="../articles/PscanR_introduction.html">Introduction to PscanR</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Federico77z/PscanR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Build background matrices in parallel</h1>
    <small class="dont-index">Source: <a href="https://github.com/Federico77z/PscanR/blob/HEAD/R/Build_Background.R" class="external-link"><code>R/Build_Background.R</code></a></small>
    <div class="hidden name"><code>ps_build_bg.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function generates a background probability profile matrix list,
used to assess if a transcription factor binding motif is appearing more
often than by chance in the pscan() input sequences, using the BiocParallel
framework for parallel execution.
It computes background scores for each motif matrix, collected, for example,
from the JASPAR database, by scanning them against a set of regulatory
sequences (e.g., gene promoters) using the `ps_scan` function.
Note that the promoter region analysed may vary.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ps_build_bg</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">pfms</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">bpparam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  BPOPTIONS <span class="op">=</span> <span class="fu">bpoptions</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  fullBG <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A `DNAStringSet` object (see Biostrings package) containing the set
of all the regulatory sequences of the organism of study
(e.g., gene promoters retrieved from a `TxDb` object).
These sequences are the target for background scanning.</p></dd>


<dt id="arg-pfms">pfms<a class="anchor" aria-label="anchor" href="#arg-pfms"></a></dt>
<dd><p>A `PFMatrixlist` object containing position frequency matrices
(PFM) representing transcription factor binding preferences.
Those are sourced from databases such as JASPAR.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>Parallelization parameter passed to `bplapply` function from
the `BiocParallel` package. This parameter defines the parallel processing
settings, including the number of cores or workers that the function uses
for computation.
See `BiocParallel` package for more details.</p></dd>


<dt id="arg-bpoptions">BPOPTIONS<a class="anchor" aria-label="anchor" href="#arg-bpoptions"></a></dt>
<dd><p>Optional configuration settings passed to `bplapply`
function from the `BiocParallel` package. This can include additional
options to control the behavior of parallel execution.
See `BiocParallel` documentation for more details.</p></dd>


<dt id="arg-fullbg">fullBG<a class="anchor" aria-label="anchor" href="#arg-fullbg"></a></dt>
<dd><p>Logical. Default is FALSE. When set to TRUE, it creates a
mapping between all sequence names in the organism of study
and the corresponding names retained after applying the unique() function.
For example, if multiple identical sequences exist (e.g., ID1, ID2, ID3,
and ID4), and unique() retains only ID2, the mapping will associate each
original name with its unique counterpart (ID1 → ID2, ID2 → ID2,
ID3 → ID2, ID4 → ID2). This vector is stored in the transcriptIDLegend
slot of the PSMatrixList and helps generate a complete background
PSMatrixList, improving computational efficiency for the pscan_fullBG()
function.
In addition, it retrieves for each PSM in the PSMatrixList output all the
background metrics relative to hits score, position, strand, and
oligonucleotide sequence of the regulatory sequences scanned with the PWM.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A `PSMatrixList` object, containing each motif matrix from `pfms`,
background-scored against the sequences in `x`.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function validates input types and removes duplicated sequences
from `x` to avoid redundant computations. It also remove all the sequences
with an N content above 50<!-- %. -->
The motif matrices are background scored by the `ps_scan` function in
parallel.
The output of this function can be stored in a .txt file by the
`ps_write_bg_to_file()` function. In this case, only information about the
background mean and standard deviation are stored for each matrix. These
metrics will be later used for the computation of z-score by the pscan
function.</p>
<p>If a full background PSMatrixList is required, that includes all the
background scores for each oligonucleotide hits, their position, strand and
names, the user should store the output by the save() function and set
the fullBG flag as TRUE. Note that this process will generate a very large
file, that can reach several gigabytes in size.</p>
<p>Note: Currently, it is not possible to use this function with the
Arabidopsis thaliana organism due to the transcript nomenclature used.
The function is designed to handle transcript names without version
extensions (e.g., NM_30287 instead of NM_30287.1).
However, in Arabidopsis thaliana, the format typically follows a
gene.transcript pattern (e.g., AT1G01010.1).
This causes the function to truncate the transcript identifier, effectively
removing essential transcript-level information. As a result, the function
may return incorrect results for this organism. This is valid for any
organism for which this type of format is used.</p>
<p>This function uses example datasets located in the `extdata/` directory for
demonstration purposes only. These files are not part of the core data used
by the function. They can be accessed using `system.file()` as shown in the
examples.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="pscan_fullBG.html">pscan_fullBG</a></code>, <code><a href="ps_write_bg_to_file.html">ps_write_bg_to_file</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load the example dataset for promoter sequences (hg38 assembly, </span></span></span>
<span class="r-in"><span><span class="co"># -200 +50 bp in respect to the TSS).</span></span></span>
<span class="r-in"><span><span class="va">file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"prom_seq.rds"</span>, package <span class="op">=</span> <span class="st">"PscanR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">prom_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">file_path</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">prom_seq</span> <span class="op">&lt;-</span> <span class="va">prom_seq</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Load the example dataset for JASPAR2020 matrices collection for </span></span></span>
<span class="r-in"><span><span class="co"># vertebrates.</span></span></span>
<span class="r-in"><span><span class="va">J2020_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"J2020.rds"</span>, package <span class="op">=</span> <span class="st">"PscanR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">J2020</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">J2020_path</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Generate the background-scored motif matrices</span></span></span>
<span class="r-in"><span><span class="va">bg_matrices</span> <span class="op">&lt;-</span> <span class="fu">ps_build_bg</span><span class="op">(</span><span class="va">prom_seq</span>, <span class="va">J2020</span>, </span></span>
<span class="r-in"><span>                           BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Use BiocParallel::MulticoreParam() for Unix like systems. </span></span></span>
<span class="r-in"><span><span class="va">bg_matrices</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> PSMatrixList of length 746</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> names(746): MA0004.1 MA0006.1 MA0019.1 MA0029.1 ... MA0748.2 MA0528.2 MA0609.2</span>
<span class="r-in"><span><span class="va">bg_matrices</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class PSMatrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ID: MA0004.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Name: Arnt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Matrix Class: Basic helix-loop-helix factors (bHLH)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> strand: +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Tags: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $alias</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "HIF-1beta,bHLHe2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $description</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "aryl hydrocarbon receptor nuclear translocator"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $family</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "PAS domain factors"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $medline</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "7592839"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $remap_tf_name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ARNT"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $symbol</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ARNT"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $tax_group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "vertebrates"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $tfbs_shape_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "11"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "SELEX"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $unibind</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $collection</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "CORE"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $species</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          10090 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "Mus musculus" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $acc</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "P53762"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Background: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    A    C    G    T </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.25 0.25 0.25 0.25 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Matrix: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         [,1]        [,2]        [,3]        [,4]        [,5]        [,6]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> A -1.5998685 -0.08004271 -4.64439090 -4.64439090 -4.64439090 -4.64439090</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> C -0.2499417 -4.64439090 -0.02927038 -4.64439090 -4.64439090 -4.64439090</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> G -4.6443909 -2.85263143 -4.64439090 -0.02927038 -4.64439090 -0.02927038</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> T -4.6443909 -4.64439090 -4.64439090 -4.64439090 -0.02927038 -4.64439090</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Background Average: 0.8205067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Foreground (your sample) Average: NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Backgroun Stdev: 0.08426005</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Background Size: 10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Foreground (your sample) Size: NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Zscore: NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan pvalue (z-test): NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example for full-background generation</span></span></span>
<span class="r-in"><span><span class="va">full_bg_matrices</span> <span class="op">&lt;-</span> <span class="fu">ps_build_bg</span><span class="op">(</span><span class="va">prom_seq</span>, <span class="va">J2020</span>, </span></span>
<span class="r-in"><span>                           BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>                           fullBG <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span>                           </span></span>
<span class="r-in"><span><span class="va">full_bg_matrices</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class PSMatrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ID: MA0004.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Name: Arnt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Matrix Class: Basic helix-loop-helix factors (bHLH)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> strand: +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Tags: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $alias</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "HIF-1beta,bHLHe2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $description</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "aryl hydrocarbon receptor nuclear translocator"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $family</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "PAS domain factors"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $medline</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "7592839"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $remap_tf_name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ARNT"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $symbol</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ARNT"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $tax_group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "vertebrates"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $tfbs_shape_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "11"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "SELEX"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $unibind</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $collection</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "CORE"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $species</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          10090 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "Mus musculus" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $acc</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "P53762"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Background: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    A    C    G    T </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.25 0.25 0.25 0.25 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Matrix: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         [,1]        [,2]        [,3]        [,4]        [,5]        [,6]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> A -1.5998685 -0.08004271 -4.64439090 -4.64439090 -4.64439090 -4.64439090</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> C -0.2499417 -4.64439090 -0.02927038 -4.64439090 -4.64439090 -4.64439090</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> G -4.6443909 -2.85263143 -4.64439090 -0.02927038 -4.64439090 -0.02927038</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> T -4.6443909 -4.64439090 -4.64439090 -4.64439090 -0.02927038 -4.64439090</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Background Average: 0.8205067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Foreground (your sample) Average: NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Backgroun Stdev: 0.08426005</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Background Size: 10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Foreground (your sample) Size: NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan Zscore: NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pscan pvalue (z-test): NA</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Federico Zambelli, Giulio Pavesi.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

